editoast/src/views/timetable/train_schedule.rs

```diff
@@ -590,10 +608,344 @@ async fn occupancy_blocks(
+    path_item_positions: &[u64],
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
+) -> Option<(String, TrackOccupancy)> {
+    let schedule_item = train_schedule
+        .schedule
+        .iter()
+        .find(|schedule| schedule.at.0 == path_item_id);
+    // Stop time at the operational point:
+    // Retrieve the stop time if it exists; otherwise, default to 0
+    let stop_duration = match schedule_item {
+        Some(schedule_item) => schedule_item.stop_for.clone().unwrap_or_default(),
+        // When the path item is the departure position without being a schedule point
+        None => PositiveDuration::default(),
+    };
+
+    // Get the index of the path item in the list of path items that corresponds to the operational point
+    let index = train_schedule
+        .path
+        .iter()
+        .position(|path| path.id.0 == path_item_id)
+        .expect("Unexcpected path item");
+    let position = path_item_positions[index];
+    let time = report_train.path_item_times[index];
+
+    // Retrieve the track section that matches the path item's position along the path
+    // TODO : Rajouter doc
+    let track_section = match path_projection.get_location(position) {
+        TrackLocationFromPath::One(track_offset) => operational_point_track_offsets
+            .iter()
+            .find(|to| *to == &track_offset)
+            .expect("Invalid track offset")
+            .track
+            .clone(),
+        TrackLocationFromPath::Two(track_offset, track_offset1) => {
+            // TODO: add assert
+            operational_point_track_offsets
+                .iter()
+                .find(|to| *to == &track_offset)
+                .or(operational_point_track_offsets
+                    .iter()
+                    .find(|to| *to == &track_offset1))
+                .expect("Invalid track offset")
+                .track
+                .clone()
+        }
+    };
+
+    Some((
+        track_section.to_string(),
+        TrackOccupancy {
+            time_begin: train_schedule.start_time + Duration::milliseconds(time as i64),
+            duration: stop_duration,
+            train_schedule_id: train_schedule.id,
+        },
+    ))
+}
+
+fn compute_track_occupancy_without_path_item_id(
+    train_schedule: &models::TrainSchedule,
+    path_projection: &PathProjection,
+    operational_point_track_offsets: &[TrackOffset],
+    report_train: &ReportTrain,
+) -> Option<(String, TrackOccupancy)> {
+    // Iterate over the track offsets linked to the operational point
+    // until we find its corresponding position on the pathfinding route.
+    // Then, retrieve the associated time (via interpolation) and position on the path.
+    for track_offset in operational_point_track_offsets {
+        if let Some(position) = path_projection.get_position(track_offset) {
+            let index = find_index_upper(&report_train.positions, position);
```

Comment:
Can this function return zero (I don't see a reason why not)? Because it will make the code below panic if this is the case.

Note: there might be an easy way out, if the index is zero, then don't interpolate and return the position and times at the index zero 🤷